woocommerce/woocommerce
# POC: Validation package

### Changes proposed in this Pull Request:

This PR is a POC around creating our own library for validation using JSON Schema.  This package was created as an alternative to using an existing library because:

* The existing libraries ([AJV](https://bundlephobia.com/package/ajv@8.17.1)) are larger in size and it's likely this library will be used on the frontend of the site (111kb vs < 10kb)
* The `$data` pointer property is needed for comparison between two properties
* Higher level of extensibility within the validation itself

<img width="695" alt="Screenshot 2024-08-14 at 11 16 31 AM" src="https://github.com/user-attachments/assets/e0ed5ef5-730b-43b0-97e2-f38c6136760e">


### How to test the changes in this Pull Request:

Since this is a POC, it's not important that this is fully functional, but this can act as a general guide for the person who takes over validation.  Here are some questions to think about:

1. Do the general concepts and exposed functions look reasonable in the [README](https://github.com/woocommerce/woocommerce/blob/add/validation-package/packages/js/validation/README.md)?
2. Is the code organized and easy to understand?  Does it scale well with the addition of new keywords and property types?
3. Could this package be moved upstream to Gutenberg?
4. Outside of the additional types and keywords needed, is there additional functionality we should consider introducing?

Some of the validation in this PR is demoed in the new product editor.  You can test this by:

1. Enabling the new product editing experience under `admin.php?page=wc-settings&tab=advanced&section=features`
2. Navigating to Products -> Add new
3. Type in the name field, see errors in the console if name is <5 characters in length
4. Note the `external_url` error in the console if not yet inserted
5. Add a space to the SKU field, note the error in the console



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
